File: modules/orientation/android/src/main/java/expo/modules/orientation/OrientationModule.kt
package expo.modules.orientation

import android.content.res.Configuration
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class OrientationModule : Module() {
  private var listener: android.content.ComponentCallbacks? = null

  override fun definition() = ModuleDefinition {
    Name("Orientation")
    Events("onChange")

    Function("getOrientation") {
      getOrientationString(appContext.reactContext?.resources?.configuration?.orientation)
    }

    OnStartObserving {
      val reactContext = appContext.reactContext ?: return@OnStartObserving
      listener = object : android.content.ComponentCallbacks {
        override fun onConfigurationChanged(newConfig: Configuration) {
          sendEvent("onChange", mapOf("value" to getOrientationString(newConfig.orientation)))
        }
        override fun onLowMemory() {}
      }.also { reactContext.registerComponentCallbacks(it) }
    }

    OnStopObserving {
      listener?.let { appContext.reactContext?.unregisterComponentCallbacks(it) }
      listener = null
    }
  }

  private fun getOrientationString(orientation: Int?) = when (orientation) {
    1 -> "portrait"
    2 -> "landscape"
    else -> "unknown"
  }
}

File: modules/orientation/ios/OrientationModule.swift
import ExpoModulesCore

public class OrientationModule: Module {
  private var orientationObserver: NSObjectProtocol?
  
  public func definition() -> ModuleDefinition {
    Name("Orientation")

    Events("onChange")

    Function("getOrientation") {
      let orientation = UIDevice.current.orientation
      switch orientation {
      case .portrait, .portraitUpsideDown:
        return "portrait"
      case .landscapeLeft, .landscapeRight:
        return "landscape"
      default:
        return "unknown"
      }
    }
    
    OnStartObserving {
      self.startOrientationObserver()
    }
    
    OnStopObserving {
      self.stopOrientationObserver()
    }
  }
  
  private func startOrientationObserver() {
    UIDevice.current.beginGeneratingDeviceOrientationNotifications()
    
    orientationObserver = NotificationCenter.default.addObserver(
      forName: UIDevice.orientationDidChangeNotification,
      object: UIDevice.current,
      queue: .main
    ) { [weak self] _ in
      let orientation = UIDevice.current.orientation
      let orientationString: String
      switch orientation {
      case .portrait, .portraitUpsideDown:
        orientationString = "portrait"
      case .landscapeLeft, .landscapeRight:
        orientationString = "landscape"
      default:
        orientationString = "unknown"
      }
      self?.sendEvent("onChange", ["value": orientationString])
    }
  }
  
  private func stopOrientationObserver() {
    if let observer = orientationObserver {
      NotificationCenter.default.removeObserver(observer)
      orientationObserver = nil
    }
    UIDevice.current.endGeneratingDeviceOrientationNotifications()
  }
}